File: frontend/observability.js
/**
 * 可觀測性統一初始化模塊 - Node.js 前端
 *
 * 負責：
 * 1. OpenTelemetry 追蹤初始化（→ 已有 Collector）
 * 2. Prometheus 指標採集（→ 已有 Prometheus via ServiceMonitor）
 * 3. 結構化 JSON 日誌（→ Loki）
 */
"use strict";

const os = require("os");
const opentelemetry = require("@opentelemetry/api");

// ─────────────────────────────────────────────────────────────
// 1. 結構化 JSON 日誌（接入 Loki）
// ─────────────────────────────────────────────────────────────
const winston = require("winston");

const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || "info",
  format: winston.format.combine(
    winston.format.timestamp({ format: "YYYY-MM-DDTHH:mm:ss.SSSZ" }),
    winston.format.errors({ stack: true }),
    winston.format.json()
  ),
  defaultMeta: {
    service: process.env.OTEL_SERVICE_NAME || "frontend",
  },
  transports: [
    new winston.transports.Console({
      format: winston.format.combine(
        winston.format.timestamp({ format: "YYYY-MM-DDTHH:mm:ss.SSSZ" }),
        winston.format.json()
      ),
    }),
  ],
});

// 將 trace_id 注入到日誌條目中
const originalLog = logger.log.bind(logger);
logger.log = function (level, message, meta = {}) {
  const span = opentelemetry.trace.getActiveSpan();
  if (span) {
    const ctx = span.spanContext();
    if (ctx && ctx.traceId) {
      meta.trace_id = ctx.traceId;
      meta.span_id = ctx.spanId;
    }
  }
  return originalLog(level, message, meta);
};

// ─────────────────────────────────────────────────────────────
// 2. OpenTelemetry 追蹤初始化（接入已有 Collector）
// ─────────────────────────────────────────────────────────────

function setupOpenTelemetry() {
  // 支持通過環境變量禁用 OTel（本地測試用）
  if (process.env.DISABLE_OTEL === "true") {
    logger.info("OpenTelemetry disabled via DISABLE_OTEL=true — running without tracing");
    return null;
  }

  try {
    const { NodeSDK } = require("@opentelemetry/sdk-node");
    const { OTLPTraceExporter } = require("@opentelemetry/exporter-trace-otlp-http");
    const { ConsoleSpanExporter } = require("@opentelemetry/sdk-trace-node");
    const { Resource } = require("@opentelemetry/resources");
    const {
      SemanticResourceAttributes,
    } = require("@opentelemetry/semantic-conventions");
    const {
      getNodeAutoInstrumentations,
    } = require("@opentelemetry/auto-instrumentations-node");

    const otelEndpoint =
      process.env.OTEL_EXPORTER_OTLP_ENDPOINT || "http://localhost:4318";

    // 嘗試 OTLP 導出器，失敗則降級爲 Console
    let traceExporter;
    try {
      traceExporter = new OTLPTraceExporter({
        url: `${otelEndpoint}/v1/traces`,
        timeoutMillis: 5000,
      });
    } catch (exporterErr) {
      logger.warn("Cannot create OTLP exporter, using Console exporter for local dev", {
        error: exporterErr.message,
      });
      traceExporter = new ConsoleSpanExporter();
    }

    const sdk = new NodeSDK({
      resource: new Resource({
        [SemanticResourceAttributes.SERVICE_NAME]:
          process.env.OTEL_SERVICE_NAME || "frontend",
        [SemanticResourceAttributes.SERVICE_NAMESPACE]: "observability-demo",
        [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]:
          process.env.ENVIRONMENT || "development",
        [SemanticResourceAttributes.HOST_NAME]: os.hostname(),
      }),
      traceExporter,
      instrumentations: [
        getNodeAutoInstrumentations({
          "@opentelemetry/instrumentation-http": {
            ignoreIncomingRequestHook: (req) => {
              return (
                req.url === "/metrics" ||
                req.url === "/health" ||
                req.url === "/favicon.ico"
              );
            },
          },
        }),
      ],
    });

    // 優雅關閉
    process.on("SIGTERM", () => {
      sdk
        .shutdown()
        .then(() => logger.info("OpenTelemetry SDK shut down"))
        .catch((err) =>
          logger.error("Error shutting down OTel SDK", { error: err.message })
        )
        .finally(() => process.exit(0));
    });

    sdk.start();
    logger.info("OpenTelemetry SDK initialized", {
      endpoint: otelEndpoint,
      exporter: traceExporter.constructor.name,
      serviceName: process.env.OTEL_SERVICE_NAME || "frontend",
    });

    return sdk;
  } catch (err) {
    logger.warn("Failed to initialize OpenTelemetry SDK, running without tracing", {
      error: err.message,
    });
    return null;
  }
}

// ─────────────────────────────────────────────────────────────
// 3. Prometheus 指標初始化（接入已有 Prometheus）
// ─────────────────────────────────────────────────────────────

function setupPrometheus(app) {
  const promClient = require("prom-client");

  // 創建 Registry
  const register = new promClient.Registry();
  register.setDefaultLabels({
    app: "observability-demo-frontend",
    service: "frontend",
  });

  // 默認指標（CPU、內存等）
  promClient.collectDefaultMetrics({ register });

  // 自定義 HTTP 指標
  const httpRequestDuration = new promClient.Histogram({
    name: "frontend_http_request_duration_seconds",
    help: "Duration of HTTP requests in seconds",
    labelNames: ["method", "route", "status_code"],
    buckets: [0.01, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10],
    registers: [register],
  });

  const httpRequestsTotal = new promClient.Counter({
    name: "frontend_http_requests_total",
    help: "Total number of HTTP requests",
    labelNames: ["method", "route", "status_code"],
    registers: [register],
  });

  const pageLoadsTotal = new promClient.Counter({
    name: "frontend_page_loads_total",
    help: "Total number of page loads",
    registers: [register],
  });

  // HTTP 中間件：記錄請求指標
  app.use((req, res, next) => {
    const start = Date.now();

    // 在響應完成時記錄
    res.on("finish", () => {
      const duration = (Date.now() - start) / 1000;
      const route = req.route ? req.route.path : req.path;

      httpRequestDuration
        .labels(req.method, route, String(res.statusCode))
        .observe(duration);

      httpRequestsTotal
        .labels(req.method, route, String(res.statusCode))
        .inc();
    });

    next();
  });

  // 暴露 /metrics 端點
  app.get("/metrics", async (req, res) => {
    try {
      res.set("Content-Type", register.contentType);
      res.end(await register.metrics());
    } catch (err) {
      logger.error("Failed to expose metrics", { error: err.message });
      res.status(500).end(err.message);
    }
  });

  logger.info("Prometheus metrics initialized");

  return { register, pageLoadsTotal };
}

// ─────────────────────────────────────────────────────────────
// 導出
// ─────────────────────────────────────────────────────────────

module.exports = {
  logger,
  setupOpenTelemetry,
  setupPrometheus,
};

File: backend/app/observability.py
"""
可觀測性配置模塊

統一管理：
- 結構化 JSON 日誌（→ Loki）
- OpenTelemetry 追蹤（→ 已有 Collector / Operator 自動注入）
- Prometheus 指標（→ 已有 Prometheus via ServiceMonitor）
"""
import os
import sys
import logging
import json
from datetime import datetime, timezone


# ─────────────────────────────────────────────────────────────
# 1. 結構化 JSON 日誌（輸出到 stdout → Loki 自動采集）
# ─────────────────────────────────────────────────────────────

def setup_logging():
    """配置結構化 JSON 日誌輸出到 stdout"""
    logger = logging.getLogger("root")
    logger.setLevel(os.environ.get("LOG_LEVEL", "INFO").upper())

    # 清除已有 handlers
    logger.handlers.clear()

    handler = logging.StreamHandler(sys.stdout)

    class JSONFormatter(logging.Formatter):
        def format(self, record):
            log_entry = {
                "timestamp": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
                "level": record.levelname,
                "name": record.name,
                "message": record.getMessage(),
                "taskName": getattr(record, "taskName", None),
                "otelSpanID": getattr(record, "otelSpanID", "0"),
                "otelTraceID": getattr(record, "otelTraceID", "0"),
                "otelTraceSampled": getattr(record, "otelTraceSampled", False),
                "otelServiceName": os.environ.get("OTEL_SERVICE_NAME", "backend"),
                "service": "backend",
            }
            return json.dumps(log_entry, ensure_ascii=False)

    handler.setFormatter(JSONFormatter())
    logger.addHandler(handler)

    # 設置 uvicorn 的日誌也使用 JSON 格式
    for _log in ["uvicorn", "uvicorn.error", "uvicorn.access"]:
        _logger = logging.getLogger(_log)
        _logger.handlers.clear()
        _logger.addHandler(handler)
        _logger.propagate = False

    return logger


# ─────────────────────────────────────────────────────────────
# 2. OpenTelemetry 追蹤配置（接入已有 OTel Collector）
# ─────────────────────────────────────────────────────────────

def setup_opentelemetry():
    """
    初始化 OpenTelemetry SDK（手動模式）。
    同時配置自動埋點（FastAPI/HTTP/日誌）和 OTLP 導出器。

    K8s 環境通過 OTEL_EXPORTER_OTLP_ENDPOINT 配置 Collector。
    本地測試可設置 DISABLE_OTEL=true 跳過。
    """
    import logging
    _logger = logging.getLogger(__name__)

    from opentelemetry import trace as otel_trace

    # 支持通過環境變量禁用 OTel（本地快速測試用）
    if os.environ.get("DISABLE_OTEL", "").lower() == "true":
        _logger.info("OpenTelemetry disabled via DISABLE_OTEL=true")
        return otel_trace.get_tracer(__name__)

    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
    from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_NAMESPACE
    from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter

    service_name = os.environ.get("OTEL_SERVICE_NAME", "backend")
    otel_endpoint = os.environ.get(
        "OTEL_EXPORTER_OTLP_ENDPOINT",
        "http://localhost:4318"
    )

    resource = Resource.create({
        SERVICE_NAME: service_name,
        SERVICE_NAMESPACE: "observability-demo",
        "deployment.environment": os.environ.get("ENVIRONMENT", "development"),
    })

    try:
        tracer_provider = TracerProvider(resource=resource)
        exporter = OTLPSpanExporter(endpoint=otel_endpoint, timeout=5)
        tracer_provider.add_span_processor(BatchSpanProcessor(exporter))
        otel_trace.set_tracer_provider(tracer_provider)
        _logger.info("OpenTelemetry SDK initialized -> %s", otel_endpoint)
    except Exception as e:
        _logger.warning("OTel Collector unreachable, using Console exporter: %s", e)
        tracer_provider = TracerProvider(resource=resource)
        tracer_provider.add_span_processor(BatchSpanProcessor(ConsoleSpanExporter()))
        otel_trace.set_tracer_provider(tracer_provider)

    return otel_trace.get_tracer(__name__)


# ─────────────────────────────────────────────────────────────
# 3. Prometheus 指標配置（接入已有 Prometheus）
# ─────────────────────────────────────────────────────────────

def setup_http_metrics():
    """註冊 HTTP 請求級別指標（替代 prometheus_fastapi_instrumentator）"""
    from prometheus_client import Counter, Histogram

    return {
        "http_requests_total": Counter(
            "http_requests_total",
            "Total HTTP requests",
            ["method", "handler", "status"],
        ),
        "http_request_duration_seconds": Histogram(
            "http_request_duration_seconds",
            "HTTP request duration in seconds",
            ["method", "handler"],
            buckets=[0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5],
        ),
    }


def setup_prometheus():
    """
    配置 Prometheus 指標採集。

    註冊自定義業務指標，返回 metrics 字典。
    /metrics 端點由 main.py 中的原生 prometheus_client 處理。
    """
    from prometheus_client import Counter, Gauge, Histogram, generate_latest, REGISTRY

    # 自定義業務指標
    metrics = {
        # CRUD 操作計數
        "todo_operations_total": Counter(
            "todo_operations_total",
            "Total number of Todo CRUD operations",
            ["operation"],  # create, read, update, delete
            registry=REGISTRY,
        ),
        # 按狀態統計 Todo 數量
        "todo_items_by_status": Gauge(
            "todo_items_by_status",
            "Number of Todo items by status",
            ["status"],  # pending, in_progress, completed
            registry=REGISTRY,
        ),
        # 按優先級統計 Todo 數量
        "todo_items_by_priority": Gauge(
            "todo_items_by_priority",
            "Number of Todo items by priority",
            ["priority"],  # low, medium, high
            registry=REGISTRY,
        ),
        # 總 Todo 數量
        "todo_items_total": Gauge(
            "todo_items_total",
            "Total number of Todo items",
            registry=REGISTRY,
        ),
        # CRUD 操作延遲
        "todo_operation_duration_seconds": Histogram(
            "todo_operation_duration_seconds",
            "Duration of Todo CRUD operations in seconds",
            ["operation"],
            buckets=[0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0],
            registry=REGISTRY,
        ),
    }

    return metrics


def update_todo_metrics(metrics, db):
    """更新業務指標（統計各狀態/優先級的 Todo 數量）"""
    from app.models import TodoItem
    from app.database import SessionLocal

    items = db.query(TodoItem).all()

    # Reset gauges
    for status in ["pending", "in_progress", "completed"]:
        metrics["todo_items_by_status"].labels(status=status).set(0)
    for priority in ["low", "medium", "high"]:
        metrics["todo_items_by_priority"].labels(priority=priority).set(0)

    # 統計
    for item in items:
        metrics["todo_items_by_status"].labels(status=item.status.value).inc()
        metrics["todo_items_by_priority"].labels(priority=item.priority.value).inc()

    metrics["todo_items_total"].set(len(items))
